adds episode discussing how to enable reproducible use of containers (#31)
carpentries-incubator/docker-introduction: _episodes/08-reproducibility.md

> ## Work in progress...
> Note that reproducibility aspects of software and containers are an active area of research, discussion and development so are subject to many changes. We will present some ideas and approaches here but best practices will likely evolve in the near future.
{: .callout}

## Reproducibility

By *reproducibility* here we mean the ability of someone else (or your future self) being able to reproduce what you did computationally at a particular time (be this in research, analysis or something else) as closely as possible even if you do not have access to exactly the same hardware resources that you had when you did the original work.

Hopefully it is reasonably self-evident why containers are an attractive technology to help with reproducibility:

  - The same computational work can be run across multiple different technologies seamlessly (e.g. Windows, macOS, Linux).
  - You can save the exact process that you used for your computational work (rather than relying on potentially incomplete notes).
  - You can save the exact versions of software and their dependencies in the image.

>   - You can access legacy versions of software and underlying dependencies which may not be generally available any more.

Reviewer: Maybe we should add another reason below this line about also keeping a copy of the data with the scripts and analysis in an image?

Zenodo supports the archiving of *tar* archives and we can capture our Docker images as tar archives using the `docker save` command:

~~~
docker save image/name:tag -o image-archive.tar
~~~
{: .bash}

These tar images can become quite large and Zenodo supports uploads up to 50GB so you may need to compress your archive to make it fit on Zenodo using a tool such as gzip (or zip):

~~~
gzip image-archive.tar
~~~
{: .bash}

Once you have your archive, you can [deposit it on Zenodo](https://zenodo.org/deposit/) and this will:

   - Create a long-term archive snapshot of your Docker image which people (including your future self) can download and reuse or reproduce your work.
   - Create a persistent DOI (*Digital Object Identifier*) that you can cite in any publications or outputs to enable reproducibility and recognition of your work.

In addition to the archive file itself, the deposit process will ask you to provide some basic metadata to classify the image and the associated work.

> Note that Zenodo is not the only option for archiving and generating persistent DOIs for images. There are other services out there - for example, some organizations may provide their own, equivalent, service.

Reviewer: I love the addition of this section.  Should we write it for them to actually do this process with an image created earlier in the lesson?

   - When you publish work (in whatever way) use an archiving and DOI service such as Zenodo to make sure your image is captured as it was used for the work and that is obtains a persistent DOI to allow it to be cited and referenced properly.

## Container Granularity

As mentioned above, one of the decisions you may need to make when containerising your research workflows
is what level of *granularity* you wish to employ. The two extremes of this decision could be characterised
as:

  - Create a single container image with all the tools you require for your research or analysis workflow
  - Create many container images each running a single command (or step) of the workflow and use them in
    sequence

Of course, many real applications will sit somewhere between these two extremes.

> ## Positives and negatives
> What are the advantages and disadvantages of the two approaches to container granularity for research
> workflows described above? Think about this
> and write a few bullet points for advantages and disadvantages for each approach in the course Etherpad.

> {: .challenge}

Reviewer: Would be good to have an answer for this challenge, even if it is exhaustive of all the possible answers learners might come up with.

### Container Orchestration

> Although you can certainly manage research workflows that use multiple containers manually, there are a number of 
> container orchestration tools that you may find useful when managing workflows that use multiple containers. 
> Introducing the use of these orchestration tools is beyond the scope of this lesson so we briefly describe 
> a few options and point to useful resources on using these tools to allow you to explore them yourself. We
> cover:

Reviewer: ```suggestion
Although you can certainly manage research workflows that use multiple containers manually, there are a number of 
container orchestration tools that you may find useful when managing workflows that use multiple containers. 
We won't go in depth on using these tools in this lesson but instead briefly describe 
a few options and point to useful resources on using these tools to allow you to explore them yourself. :
```

- "Using a service such as Docker Hub allows us to easily share computational work we have done."
- "Using container images along with a DOI service such as Zenodo allows us to capture our work and enables reproducibility."
- "Tools such as Docker Compose, Docker Swarm and Kubernetes allow us to describe how multiple containers work together."
---

Although this workshop is titled "Reproducible computational environments using containers",
so far we have mostly covered the mechanics of using Docker with only passing reference to
the reproducibility aspects. In this section, we discuss these aspects in more detail.

> ## Work in progress...
> Note that reproducibility aspects of software and containers are an active area of research, discussion and development so are subject to many changes. We will present some ideas and approaches here but best practices will likely evolve in the near future.
{: .callout}

## Reproducibility

By *reproducibility* here we mean the ability of someone else (or your future self) being able to reproduce
what you did computationally at a particular time (be this in research, analysis or something else)

> as closely as possible even if you do not have access to exactly the same hardware resources #

Reviewer: The sentence is written as someone else (except in parentheses) so I think it should be they here. 
```suggestion
as closely as possible even if they do not have access to exactly the same hardware resources #
```
